[WIP] Created a basis on which to discuss EOM class

jbm950 opened the pull request:
This example is going to walk through the simulation of a two mass
spring damper system where the equations of motion are provided for
pydy's System by hand. This initial commit hold the base code from eom
formation to simulation and display of results.

jbm950 committed:
Created a basis on which to discuss EOM class

This example is going to walk through the simulation of a two mass
spring damper system where the equations of motion are provided for
pydy's System by hand. This initial commit hold the base code from eom
formation to simulation and display of results.

jbm950 committed:
Added drawing and readme to two mass spring damper example

I drew up the example system and created a README file. The API
discussion should be easier with the drawing to reference.

moorepants commented on examples/two_mass_spring_damper/run.py:
```diff
@@ -0,0 +1,53 @@
+
+# Initialize the equation of motion class
+
+eom = eombase.EOM(coordinates, speeds, mass_matrix, forcing)
```
This should be able to accept multiple forms of the eoms. It may be good to look at the three options available for the ODEFunctionGenerator class in PyDy for examples. We may simply need to add one more option to handle DAEs.

moorepants replied:
It would also be nice to support "output equations" here, i.e. those equations that are algebraically dependent on the state. For example, kinetic energy of the system.

jbm950 replied:
I'm confused by your second comment. Are you suggesting that the class `EOM` should be able to produce the kinetic energy and such of the system?

moorepants replied:
I'm suggesting that it may be a good idea to have the EOM class contain additional arbitrary equations that are functions of the states, parameters, etc. One example of an equation such as this is an expression for kinetic energy.

moorepants replied:
The auxiliary equations in KanesMethod are also examples.

jbm950 committed:
Added additional methods of manual eom entry

Added demonstrations of entering the equations of motion manually into
the eom class. The README was also changed to reflect this alteration in
the goal of the example.

jbm950 committed:
Added a portion showing the ability to add equations to EOM

Added a demonstration to the code to show the ability to add equations
to the EOM class to be stored for later access. The output_eqns keyword
should allow multiple additional equations to be added. These equations
will be accessed by the keyword of the output_eqns dictionary.

jbm950 commented:
Was something like that generally what you had in mind?

moorepants commented on examples/two_mass_spring_damper/run.py:
```diff
@@ -0,0 +1,86 @@
+#    [3] M(q, p) u' = F(q, u, t, r, p)
+#        q' = G(q, u, t, r, p)
+
+eom1 = eombase.EOM(coordinates, speeds, rhs=RHS, output_eqns=out_eqns)
```
`rhs` can probably be a required arg since all forms of the equations have a right hand side.

jbm950 replied:
Your comment confuses me. `rhs` used in `eom1` is specifically F(x, t, r, p) from form [1]. If it were to be generally the right and side of any of the other two forms then the input would have different meanings in different situations. Also this would get rid of `forcing` and `forcing_full` keyword arguments.

moorepants replied:
All three sets of equations have a "right hand side" as shown, i.e. whatever is on the right hand side of the equals sign. Thus for every form lists, the user must always pass in a right hand side of the equations. But the coefficient matrices on the left hand side are optional.

In the ODEFunctionGenerator class I'm pretty sure it always requires the user to pass in the right hand side of whatever form of the equations they are using.

jbm950 replied:
Currently in `KanesMethod` and `LagrangesMethod` the right hand side of [1] is obtained by `*Method.rhs()`, the right hand side of [2] is obtained by `*Method.forcing_full` and the right hand side of [3] is obtained by `*Method.forcing`.

From the docstring of `ODEFunctionGenerator`
```
        right_hand_side : SymPy Matrix, shape(n, 1)
            A column vector containing the symbolic expressions for the
            right hand side of the ordinary differential equations. If the
            right hand side has been solved for symbolically then only F is
            required, see form [1]; if not then the mass matrix must also be
            supplied, see forms [2, 3].
```
I suppose the confusion is in not having a more formal name for the right hand side of [1] than simply "rhs"

moorepants replied:
Yes, the name is too general and confusing.

jbm950 replied:
Do you have any ideas on something we can change it to? Maybe something like `first_order_form_rhs` or `fof_rhs` for brevity. This is assuming that [1] is the only one represented first order form as implied by the docstring for `rhs()` in `KanesMethod`

```
    def rhs(self, inv_method=None):
        """ Returns the system's equations of motion in first order form.
```

moorepants replied:
I've been working on this stuff for about 8 years now and have yet to devise a good textual naming for all this stuff. I'm still not even sure I'm using all the "correct" terminology. Some words to throw out there:

- ordinary differential equation
- implicit ode
- explicit ode
- differential algebraic equation
- right hand side
- mass matrix
- coefficient matrix
- first order ode
- second order ode

I not sure there is a consensus on how to describe the different "forms" of these ODE/DAEs. And when you try you end up with very verbose words. Maybe check the math and dynamics literature for some ideas.

jbm950 replied:
`Maybe check the math and dynamics literature for some ideas.` 

I'm up for doing the research but I don't know where I would start. Do you have a suggestion?

moorepants commented on examples/two_mass_spring_damper/run.py:
```diff
@@ -0,0 +1,86 @@
+
+eom1 = eombase.EOM(coordinates, speeds, rhs=RHS, output_eqns=out_eqns)
+eom2 = eombase.EOM(coordinates, speeds, mass_matrix_full=mm_full,
+                   forcing_full=f_full)
```
One thing that was raised on the other pydy pr about this was whether a general EoM object would need to differentiate between the coordinates and speeds or could simple a state vector be passed in.

moorepants replied:
Wikipedia and wolfram.

I've been following kane's book to fix up the kane's notation.

I'm sure that people have variations on the lagrange notations since his stuff is so old.

For general ode stuff look up an ode book, but if you could find a popular text book that was about dynamics that covers this stuff that would be good. Maybe Greenwood?

jbm950 replied:
`self.coordinates` is only used in two places in `System`: being combined with `self.speeds` to form `self.states` and in `_args_for_gen_ode_func`. The latter of the two places I've chased down and has currently brought me to the internals of `ode_function_generators.LambdifyODEFunctionGenerator` where I am currently searching to see if the coordinates are actually needed separate from the speeds. I'll post here when I find the answer.

moorepants replied:
It probably isn't needed internally to the ode generators. But if you were to create methods on system like:

```python
sys.plot_coordinate_trajectories()
```

It would be nice to have them separate. You can work with just the state vector for visualization and further work with system but for multbody systems, more often than not, you look at the coordinates and speeds separately because they are different scales. You could even imagine how knowing the units of the states would help improve automatic viz and other things.

moorepants replied:
So I guess we need to decide if System is very general or is System made for multibody systems. That was what some of the discussion in pr #345 was about. I personally would like System to be a multibody system so that it is smart enough to do normal things you want to do with a multibody system. But a `MultibodySystem` could be a subclass to a more general `System`. Systems should also tie nicely into pydy.viz. Right now it is a bit clunky, but a System could be aware of the bodies and loads in the system and make a default visualization.

jbm950 replied:
`So I guess we need to decide if System is very general or is System made for multibody systems. `
I would say having a general class that is subclassed would provide a broader base for future functionality.

Having just a single variable `states` would give the class less information overall to work with but would be more compact. If subclasses need to know which states are coordinates and which are speeds they could have a simple input variable for the number of coordinates and assume the first elements that correspond to that number are the coordinates and the rest are speeds. 

`You could even imagine how knowing the units of the states would help improve automatic viz and other things.`

I would think this could be very simply by inputting a list of unit names to either `EOM` or `System` where they positionally correspond to coordinates/speeds in the `states` variable. This solution wouldn't require the coordinates and speeds to be separate variables. Example code
```
eom = EOM(states=[q1, q2, q2, u1, u2, u3], units=['m', 'm', 'mm', 'm/s', 'cm/s', 'mm/s'])
```

moorepants replied:
That all sounds fine about system being general.

For units, we should think about how we may utilize the units functionality already available in sympy.

moorepants commented:
> Was something like that generally what you had in mind?

Yes, now we need to think about how to incorporate the DAE forms from Lagrange's (or other) methods. And think about what are the attributes and methods of EoM.

jbm950 committed:
Added an example problem with DAEs

Added an example of the simple pendulum being defined by x and y
coordinates such that it resulted in a DAE. This system is passed
manually to eombase class.

jbm950 committed:
changed some of the API suggestions in the two mass spring damper system

Added the ability to pass in either a combination of coordinates and
speeds or just a state vector. Also altered how the output equations are
accessed and how they are used.

jbm950 committed:
Added pydy.system use to simple pendulum example

Added some code to the simple pendulum example to include a time
simulation and some basic output from the system class

jbm950 committed:
added example of System plotting coordinates

Added an example api of system plotting coordinates from manual eombase
input and from lagranges equation eom formation.

jbm950 commented:
I've begun adding other examples for the API of `eombase`. One addition is the ability to pass in coordinates and speeds together or a single states iterable.

moorepants commented on examples/simple_pendulum/run.py:
```diff
@@ -39,3 +40,21 @@
+
+# Display the kinetic energy change in time (obtained from the particle in the
+# bodies list)
+KE = sys.body_kinetic_energies()
```
What will KE be? Is this similar to out? i.e. an array of trajectories for kinetic energy?

jbm950 replied:
I think that would be a good idea to have the same output. sys.integrate() seems to give the trajectories such that each column represents a single state. I feel this could be put to the same use here where each column represents a single body.

moorepants replied:
Same style of output sounds good. But what are "kinetic energies"? Each body/particle has kinetic energy and then the whole system has kinetic energy, as does any combination of bodies and particles. This needs clearer definition.

jbm950 replied:
I was imagining it being the kinetic energy of each individual body since you said that a benefit of the `*Methods` is that they contain extra info such as a list of each of the bodies which in turn keeps track of its kinetic energy.

moorepants replied:
Ok, that could work. You need to then have a way for the user to know which body/particle that the rows/columns of the array correspond to. And the kinetic energy of the whole system is important too. Would the user just sum along the axes of the numpy array for the total kinetic energy or would you precalculate that an it would be an extra row/col? This is where it would be good to show the desired output.

moorepants commented on examples/simple_pendulum/run2.py:
```diff
@@ -0,0 +1,55 @@
+G = Matrix([u, v])
+
+# Form the rhs of the dynamics equations
+RHS = mm_full.inv()*f_full
```
Typically best to use `mm_full.LUsolve(f_full)` instead of explicit inversion.

moorepants commented on examples/simple_pendulum/run2.py:
```diff
@@ -0,0 +1,55 @@
+eom2 = eombase.EOM(states=states, mass_matrix_full=mm_full, forcing_full=f_full,
+                   alg_con=alg_con_full)
+eom3 = eombase.EOM(states=states, mass_matrix=mm, forcing=f, kinematics=G,
+                   alg_con=alg_con)
```
These look good. Some comments on args versus kwargs. Some of these things should probably be args instead of kwargs. For example, each system requires that there be coordinates and speeds OR states. So:

```python
EOM(coordinates, speeds, ...)
EOM(states, ...)
```

Seems more natural and avoids using the kwargs.

Also, rhs is always required. So

```python
EOM(states, rhs, mass_matrix=M)
```

If you look carefully at the API here:

https://github.com/pydy/pydy/blob/master/pydy/codegen/ode_function_generators.py#L186

You'll see that it has a set of required args and a set of optional args. It tries to minimize the total number of distinct args, e.g. only mass_matrix not mass_matrix and mass_matrix_full. This then relies on a good docstring to show how to pass in the different combinations of args/kwargs. I prefer this approach more than making a ton of only optional kwargs (I'm biased though, as I wrote the codegen api).

moorepants replied:
In fact, is there any reason not to copy the codegen api so that we have consistency throughout the code?

jbm950 replied:
`"For example, each system requires that there be coordinates and speeds OR states."`
I was thinking about allowing the user to pass in either of the two in which case I think kargs would make more sense. In the very least coordinates/states could be a single argument and speeds could be a karg. The class could then test input format by seeing if speeds was passed. By making them all arg inputs could cause errors when using a just states. Take the code snippet you posted later in the changes for example.

```
eom1 = eombase.EOM(coordinates, speeds, RHS, output_eqns=out_eqns)
eom2 = eombase.EOM(states, forcing_full, mass_matrix=mm_full, num_coordinates=2)
eom3 = eombase.EOM(states, f, mass_matrix=mm, kinematics=G)
```

In eom2, forcing_full would be passed to the speeds variable if the class was set up for `__init__(coordinates/states, speeds, forcing/rhs, ...)` If it is decided to only use a single mode of input I believe using `states` would be the better option.

(writing response to second question now)

moorepants replied:
You can use `__init__(*args, **kwargs)` and parse the input in any fashion you want to.

moorepants replied:
We do this in the VisualizationFrame class (probably too much).

You could do this:

```python
eom1 = eombase.EOM(coordinates, RHS, speeds=speeds, output_eqns=out_eqns)
eom2 = eombase.EOM(states, forcing_full, mass_matrix=mm_full, num_coordinates=2)
eom3 = eombase.EOM(states, f, mass_matrix=mm, kinematics=G)
```

jbm950 replied:
`In fact, is there any reason not to copy the codegen api so that we have consistency throughout the code?`

Currently the `__init__` for ODEFunctionGenerator is
```
    def __init__(self, right_hand_side, coordinates, speeds, constants,
                 mass_matrix=None, coordinate_derivatives=None,
                 specifieds=None, linear_sys_solver='numpy',
                 constants_arg_type=None, specifieds_arg_type=None):
```
Coordinates and speeds could be moved to *args like you suggested. I think the key inputs here would then be `right_hand_side, *args, mass_matrix=None, coordinate_derivatives=None` not including the other arguments we've discussed elsewhere in the PR (like `output_eqns`)

jbm950 replied:
```
eom1 = eombase.EOM(coordinates, RHS, speeds=speeds, output_eqns=out_eqns)
eom2 = eombase.EOM(states, forcing_full, mass_matrix=mm_full, num_coordinates=2)
eom3 = eombase.EOM(states, f, mass_matrix=mm, kinematics=G)
```
This is what I was imagining where the speeds karg lets the class know explicitly the form the user input.

jbm950 replied:
In keeping with the codegen api though the argument `kinematics` would need to be changed to `coordinate_derivatives`

moorepants commented on examples/two_mass_spring_damper/run.py:
```diff
@@ -0,0 +1,94 @@
+
+# Form the rhs of q' = G(q, u, t, r, p). Kinematic equation
+
+G = Matrix([u1, u2])
```
I'm pretty sure that lagrange's method spits out q, q', and q'', which are not in explicit first order form, i.e. q'=u hasn't been used to substitute the u's for the q'. This means the lagrange's method gives the equations in second order form instead of first order form, right? How will this class deal with that?

jbm950 replied:
I agree that lagrange's method uses q, q' and q'' and doesn't even store a "speeds" variable (`LagranesMethod.u` returns `LagrnagesMethod._qdots`). This results in `Derivative(q(t), t)` potentially showing up in `L.mass_matrix` and `L.forcing`. I suppose these would need to be replaced with `u`'s so that no derivative terms show up in either `L.mass_matrix` or `L.forcing`. Leaving the equations in terms of q'' does not have any other functional difference that I can think of because `LagrangesMethod.mass_matrix_full` and `LagrangesMethod.forcing_full` specify that u=q' and u' or q'' doesn't actually show up in the resulting symbolics.

moorepants replied:
Good to know. We need to figure out how to handle this. Lagrange's method should have a way to covert from the second order form to first order form. Kane's method does that explicitly from the beginning by specifying the kinematic differential equations.

moorepants commented on examples/two_mass_spring_damper/run.py:
```diff
@@ -0,0 +1,94 @@
+# equation
+
+KE = 0.5 * (m1*u1**2 + m2*u2**2)
+out_eqns = {"kinetic_energy": KE}
```
Since everything is a variable in the equations of motion, this should map a symbol to an expression instead of mapping a string to an expression. That seems to be more consistent.

Also, should we allow simply an iterable of expressions too? That's the way most things are passed into the `*Method`s classes, for example.

jbm950 replied:
`Since everything is a variable in the equations of motion, this should map a symbol to an expression instead of mapping a string to an expression. That seems to be more consistent.`

Are you meaning something along the lines of

```python
KE = symbols('KE')
out_eqns = {KE: expr}
```

`Also, should we allow simply an iterable of expressions too? That's the way most things are passed into the *Methods classes, for example.`

Are you meaning

```python
[1]: out_eqns = {'Kinetic Energy': [expr1, expr2, ...]}
```

or 

```python
[2]: out_eqns = [expr1, expr2, ...]
```

The reason I chose to use a string is so each equation could be differentiated by a distinct label. This would be useful when there are a variety of types of output equations being used as each one would be able to be labeled distinctly its name of the equation (and could later be pulled by name).

moorepants replied:
Most things are passed into the `*Method`s classes as:

```python
exprs = [expr1, expr2, expr3, ...]
```

Where `expri` represents the left hand side of:

```
expri = 0
```

In the `System` class we use mostly dictionaries to map sympy symbolics to numerics (floats or arrays of floats). `System` was designed years after `*Method`s and has some nice advantages. So in `System` you could do something like:

```python
sys.plot_state(q1, q2, q3, u1)
```

Where q1, q2, q3, u1 are SymPy symbols. They can be easily printed in the plot using LaTeX. But you may also sometimes what to know what a state is, i.e. q1 is the angle angle between the arm of the pendulum and the ground...

If you use symbols for the output equations you can do:

```python
sys.plot(q1, q2, ke)
```

Where you can know include other symbols that represent the output equations.

I don't know what the best idea for all these API choices are, keep in mind I'm just throwing out ideas and trying to insert some of the historical information about the API design.

jbm950 replied:
`Most things are passed into the *Methods classes as:... expri = 0`

I think that the form `out_eqns = {'Kinetic Energy': [expr1, expr2, ...]}` is a good combination of the previous API where `expri = 0` still holds and you can differentiate between the basic equations you put in. I think `out_eqns= [expr1, expr2, expr3, ...]` would be too vague because there is a variety of different output equations you could want to specify (for example expr1 = kinetic energy of body 2, expr2 = contact forces in joint3, expr4 = auxiliary equation).

`In the System class we use mostly dictionaries ... Where you can now include other symbols that represent the output equations.`

Might it be that strings or symbols could be used for the keys in the dictionary? I think strings might be simpler because some of the output equations could be specified by the `*Methods` class in the `*Methods.to_eom()` method (like Kane's auxiliary equations) and the user would have to come up with the symbol that the method used instead of just a string with the label for the expression (which brings up the need to be able to pass additional equations to `output_eqns` outside of `EOM.__init__(...)`).

moorepants commented on examples/two_mass_spring_damper/run.py:
```diff
@@ -0,0 +1,94 @@
+                   output_eqns=out_eqns)
+eom2 = eombase.EOM(states=states, mass_matrix_full=mm_full, forcing_full=f_full,
+                   num_coordinates=2)
+eom3 = eombase.EOM(states=states, mass_matrix=mm, forcing=f, kinematics=G)
```
How about:

```python
eom1 = eombase.EOM(coordinates, speeds, RHS, output_eqns=out_eqns)
eom2 = eombase.EOM(states, forcing_full, mass_matrix=mm_full, num_coordinates=2)
eom3 = eombase.EOM(states, f, mass_matrix=mm, kinematics=G)
```

Why is num_coordinates needed? And how do you know what order the coordinates are in wrt to the state vector?

moorepants replied:
I think we need a way to pass in the bodies and loads too.

Also, can this work:

```python
EOM(kanes_method_instance)
EOM(lagranges_method_instance)
```
?

Or should this be how you do it:

```
kane = KanesMethod(...)
eom_instance = kane.to_eom()
```
?

jbm950 replied:
`Why is num_coordinates needed? And how do you know what order the coordinates are in wrt to the state vector?` 

Number of coordinates is needed for the `sys.plot_coordinates()` method that was discussed last Friday as being a potentially useful method for system. It would be assumed that state is held in the following format [coordinates, speeds, other (ex. Lagrange coefficients)] in which case knowing the number of coordinates the code can pick that many elements off the beginning of the states iterable and have just the coordinates.

`Also, can this work`

This would not be the ideal method because then EOM would have to account for and prepare for every single method that could be used. This would be a maintenance nightmare and make adding new methods an annoying difficult process which is the opposite of the desired result.

`Or should this be how you do it:`

This would be one method. In this way each `*Method` can organize its results into the format that is expected by EOM and would make adding new methods easier as they would simply form the equations of motion their own way then format them for the `EOM` class.

The other way would be `class *Method(eom_base):`. In this method the equations of motion will still need to be specified to the universal API where the base class would provide a blue print. This would require more faith in the programmer of the new `*Method` class to understand and mimic the desired API where the returned object would automatically have the desired API. 

Ex.
```python
class eom_base(object):
    @property
    def mass_matrix(self):
        return self._m_d
```
would require the designers of new classes to include `self._m_d` in their definitions. The advantages of subclassing is that the different components of `*Methods` remain with the equation of motion data (ex. `auxiliary_eqns` or `solve_multipliers()`). For our purposes though I think a returned `EOM` instance might be better as the user can still use the additional information as the `*Method` instance hasn't been deleted, it just isnt what is passed to system.

`I think we need a way to pass in the bodies and loads too.`

I agree this can be useful information.

moorepants commented on examples/two_mass_spring_damper/run.py:
```diff
@@ -0,0 +1,94 @@
+# Returns a dictionary whith the same keys as output_eqns but the values are the
+# numerical results of the simulation
+
+print(eom1.output_eqns_results)
```
What does this print?

moorepants replied:
This made me think about why we aren't using documentation, instead of just the input file, to show things. For example you could do:

```python
>>> kane = KanesMethod(...)
>>> kane.states
Matrix([q1, q2, u1, u2])
>>> eom = kane.to_eom()
>>> eom.states
Matrix([q1, q2, u1, u2])
>>> sys = System(eom)
>>> sys.integrate()
array([1.0, .2.0, 3.0, 4.0])
>>> sys.output_equation_trajectories()
array([5.0, 4.0, 6.0])
```

This shows the desired input and output, is more akin to a unit test, and is called "documentation driven development" if you write this first (as you are doing).

jbm950 replied:
`What does this print?`

I was thinking something like `{'kinetic energy': np.array([...])}`

`This shows the desired input and output,`

This is a good idea. I'll set up a page with the three examples that have been used so far.

moorepants replied:
You might as well do this in the sphinx documentation, because once everything is implemented these will simply become doctests.

jbm950 committed:
Moved examples to documentation

This documentation will contain examples of eombase and it's usage.

jbm950 commented:
I have started a documentation page where I have placed all three examples. I am currently changing them from their code form to documentation format and have then I will begin addressing the changes we discussed above.

jbm950 committed:
Finished changing examples documentation

The examples are now completely changed from script code format to a
documentation format.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,233 @@
+states without you having to call matplotlib yourself. ::
+
+    >>> sys.plot_coordinates()
+    >>> sys.plot_states()
```
How about a `plot_trajectories` method that works like so:

```python
>>> sys.plot_trajectories(*sys.coordinates)
>>> sys.plot_trajectories(sys.coordinates, sys.speeds)  # makes sub plots one for coords one for speeds
>>> sys.plot_trajectories((q1, q2), (u1, u2))  # also two subplots
>>> sys.plot_trajectories(q1, q2, ke)  # takes any symbol that system knows about, ke would have to be the key hen defining the output equations
>>> sys.plot_trajectories(c * u + k * q)  # we could even let the user plot an expression of symbols that are known to system
```

jbm950 committed:
Altered API examples in eom_base.rst

The API examples are now updated to reflect the discussion for the
development of eombase. Input to eombase.EOM has been altered, more
attribute access examples have been added and output examples have been
added.

jbm950 committed:
Added a plot_trajectories API to the example documentation

The example documentation now includes a proposed plot_trajectories api.
This will be used for the other plot methods contained within system.
Example:

    def plot_coordinates(self):
        self.plot_trajectories(self.coordinates)

jbm950 commented:
Your plot trajectories comment has been addressed along with the other items we've discussed thus far.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,255 @@
+    >>> sys.output_eqns()
+    {"kinetic energy": array([[0.0], [...], ...])}
+    >>> eom1.output_eqns_results
+    {"kinetic energy": array([[0.0], [...], ...])}
```
I think that you may be able to allow any key here, str, Symbol, etc. And then when the user asks for a specific symbol it will search the coordinates, speeds, specified, and parameters (all sympy symbols) and then finally it will look for the key in the output_equations dictionary, which can be anything. If it is a sympy symbol, the plotting will render it with LaTeX, but if it is any other arbitrary key it will print as whatever the `__str__` method of that object produces. So this if fine, but I think the default should be that you pass in a SymPy symbol, because currently System is only aware of SymPy symbols and functions. In the far future we could imagine having a specialized symbol/function that did this:

```python
>>> q = DynamicSymbol('q', name='Angle between ground and arm', unit=radian)
>>> q
q(t) rad
>>> q.name
Angle between ground and arm.
>>> q.unit
radian
>>> theta = DynamicSymbol('theta', unit=degree)
>>> q + theta
q(t) + theta(t) * pi / 180 rad
>>> ke = DynamicSymbol('K_e', name='Kinetic energy of the arm')
>>> output_equations = {ke: 1 / 2 * I * q.diff() ** 2} 
```

This would then allow us to have a rich way to print all these objects in the various visualizations. But for now we don't have anything but SymPy symbols to work with. The name and units are not yet part of the whole design, so we should keep things consistent.

jbm950 replied:
I could add a second output equation to the dictionary to show that strings and symbols are potential keys.

moorepants replied:
Sounds good, These will eventually  be translated to unit tests that will ensure it can work with a variety of inputs.

jbm950 committed:
Added to output equations in eombase docs

Added an example to the output equations to show that strings or symbols
are acceptable keys of the output_eqns dictionary.

jbm950 commented:
I have put my notes that I have regarding equations of motion "forms" that I have found so far on my [website](http://jbm950.github.io/notes/dynamics/dynamics_main.html). Most sources that I have found are only bothered with forming the equations of motion and make no mention of different arrangements and what the names of those arrangements might be. I'm not sure that there even are names for the different arrangements and we may even be free to name them ourselves for our software.

jbm950 committed:
Fixed expected output form the output equations call

Last commit I added an output equation example by I forgot to change the
output of calling for the results of the output equations. Calling the
output equations will now include the potential energy.

jbm950 committed:
Finialized initial draft of EOM class example

The examples document now has an example of every method and every
public attribute of the EOM class.

jbm950 commented:
I have written the rest of the public attributes and methods into the example. If there are any I missed or didn't think of let me know. I'm going to go through and add common error messages to specify invalid input combinations or attribute accesses when inputs aren't passed.

jbm950 committed:
Added error explainations

Went through the examples and added explainations and examples of
different types of errors that could occur when using eombase.EOM

jbm950 commented:
@angadhn or @moorepants could someone do a review of the .rst file? Also I think the check failed because of a change to numpy.

oliverlee commented:
I've merged the fix so feel free to merge/rebase this branch.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+    >>> from numpy import linspace
+    >>> from pydy.system import System
+    >>> from sympy import symbols, Matrix
+    >>> from sympy.physics.mechanics import dynamicsymbols, eombase
```
`>>> from sympy.physics.mechanics.eombase import EOM`

Symbolic Equations Of Motion: EoM, EOM, EquationsOfMotion

`*Methods` classes will generate equations of motion.

moorepants replied:
`from . import EquationsOfMotion as EOM`

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+    >>> eom1 = eombase.EOM(coordinates, RHS, speeds=speeds, 
+    ...                    output_eqns=out_eqns)
+    >>> eom2 = eombase.EOM(states, f_full, mass_matrix=mm_full,
+    ...                    num_coordinates=2)
```
I don't think `num_coordinates` is sufficient. For example, what if the states are in this order:

`q1, q2, u3, u2, q3, u1`

We should allow any order of the states, as users that manually generate them might want an arbitrary  order.

The other thing to note is that `q_i` does not necessarily correspond to `u_i` in the general sense and `len(q)!=len(u)`. So, there needs to be a way to tell the class which indices are coordinates or speeds.

Maybe `coord_idxs=(3, 2, 5, 9)`.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+input combinations, all three of the equations of motion forms will be used. ::
+
+    >>> eom1 = eombase.EOM(coordinates, RHS, speeds=speeds, 
+    ...                    output_eqns=out_eqns)
```
I like the optional speeds so that the first arg is the states.

moorepants replied:
Would 'SymbolicSystem" be a better name than `EquationsOfMotion`, being that we include bodies and loads in addition to the questions.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+    SyntaxError: Need to specify a mass matrix for eom form [3]
+    >>> eom1.mass_matrix
+    AttributeError: Mass matrix is not specified for eom form [1]
+    >>> eom1.mass_matrix_full
```
forcing ->  right_hand_side
mass_matrix -> mass_matrix

mass_matrix_full -> mass_matrix_first_order
forcing_full -> right_hand_side_first_order

mass_matrix -> dynamic_mass/inertial_matrix
kinematic_matrix
composite_mass_matrix, composite_matrix

# Dynamic Differential Equations
## Implicit
M u' = F
## Explicit:
u' = inv(M) * F

# Kinematic Differential Equations
## Implicit
H q' = G
## Explicit
q' = inv(H) * G

# Combined Differential Equations
## Implicit
W x' = Y
## Explicit
x' = inv(W) * Y

moorepants replied:
forcing ->  right_hand_side
mass_matrix -> mass_matrix

mass_matrix_full -> mass_matrix_first_order
forcing_full -> right_hand_side_first_order

mass_matrix -> dynamic_mass/inertial_matrix
kinematic_matrix
composite_mass_matrix, composite_matrix

# Dynamic Differential Equations
## Implicit
M u' = F
## Explicit:
u' = F* = inv(M) * F

# Kinematic Differential Equations
## Implicit
H q' = G
## Explicit
q' = G* = inv(H) * G

# Combined Differential Equations
## Implicit
W x' = Y
## Explicit
x' = Y* = inv(W) * Y

F, G, and Y are the right hand side of the implicit forms of the * differential equations.
M, H, and W are the (inertial) coefficient matrices of the implicit forms of the * differential equations.
F*, G*, and Y* are the right hand side of the explicit forms of the * differential equations.

moorepants replied:
F: `.dyn_implicit_rhs`
G: `.kin_implicit_rhs`
Y: `.comb_implicit_rhs`
M: `.dyn_implicit_mat`
H: `.kin_implicit_mat`
W: `.comb_implicit_mat`
F*: `dyn_explicit_rhs`
G*: `kin_explicit_rhs`
Y*: `.comb_explicit_rhs`

moorepants replied:
I wonder if we could do something like:

EquationsOfMotion.implicit.dynamic.rhs
EquationsOfMotion.implicit.dynamic.mat

or

EquationsOfMotion.explicit_dynamic().rhs

jbm950 replied:
For the inputs to the SymbolicSystem initialization I'm leaving mass_matrix and coordinate_derivatives to both match the current api and because mass_matrix can take in either dyn_implicit_mat or comb_implicit_mat. I'll change the attributes to match the new names.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+will be returned. ::
+
+    >>> eom.output_eqns_results
+    ValueError: The output equations first need to be calculated
```
Both integrate() and evaluating the output equations require computation (sometime a lot), so they should be methods. but if we want to plot the trajectories, then it would be convenient to have them stored in System. Or the functions on System like `.plot()` should take the trajectories (a numpy array) as an argument. I'm not sure what is best. Or we can return a result object that has the plot methods.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+     PE: array([[5.0], [...], ...])}
+    >>> eom1.output_eqns
+    {"kinetic energy": KE, PE: 0.5*k1*x1**2 + 0.5*k2*(x2-x1)**2}
+    >>> eom1.output_eqns_results
```
How about `output_trajectories` and `state_trajectories` for these attributes. Also, we should provide the state and output trajectories in the same form, either a dictionary mapping keys to 1d arrays, 2d arrays containing a row for each state/output, or both. The 2d array form is quite convenient for plotting purposes and more efficient gross numerical computations with numpy. If we wanted a new dependency we could return a pandas DataFrame which is like a dict but has all kinds of niceties for time series analysis.

jbm950 replied:
I do not quite understand the 2d array form you have described. Since this is more related to pydy's System object now I'm going to move forward with SymbolicSystem in SymPy.

moorepants replied:
That's fine.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+
+    >>> sys.plot_coordinates()
+    >>> sys.plot_speeds()
+    >>> sys.plot_states()
```
All above should call line 132 internally I suppose.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+    (x, y)
+    >>> eom3.speeds
+    (u, v)
+    >>> eom1.rhs
```
Is `right_hand_side` preferable.

moorepants commented on docs/eom_base.rst:
```diff
@@ -0,0 +1,372 @@
+    Matrix([[1, 0, -x/m], [0, 1, -y/m], [0, 0, l**2/m]])
+    >>> eom1.alg_con
+    [4]
+    >>> eom1.dynamic_symbols()
```
Why a method? Should it be an attribute?

moorepants commented:
@oliverlee We are working on this API here at SciPy this week. Would you have some time to comment soon? We'd like to home in on a nice API that can describe a rigid body system's symbolic equations of motion in a general way.

moorepants commented:
@jbm950 I'm happy with moving forward with this. I've made comments about things I think should be adjusted. There are some details about naming but I think moving forward with the idea about having a symoblic system description object is good. All of the Method objects can then return this object as a primary result.

jbm950 committed:
Renamed many of the items in eom_base.rst

Changed the name of eombase and EOM. Also renamed all of the attributes
dealing with the different forms of the equations of motion.

jbm950 committed:
Removed the attribute for symsystem for the output equation results

Since the class is a symbolic representation of the system I didn't
thing it made sense anymore to store the numerical results of the
integration on the object. This information is now stored in pydy's
System class.

jbm950 committed:
Changed how coordinates and speeds are specified

When just a list of the states is passed in to SymbolicSystem, the
coordinates and speeds are now determined by giving the object a list of
the indexes of the coordinates in speeds. This allows arbitrary order in
the states iterable.

moorepants commented:
This is looking good. The only thing I'm not sure is addressed is the issue associated with the fact that users may not order there states in any particular order. There was some assumption that `states = coordinates + speeds`. A user that manually populates the state vector may order them in any fashion, for example: `[u2, q3, u1, q2, q1]`. So maybe there should be two attributes that list the indices of the coordinates and speeds, in this case: `q_idxs = [4, 3, 1]` and `u_idxs = [2, 0]`.

jbm950 commented:
```python
>>> symsystem2 = system.SymbolicSystem(states, comb_implicit_rhs, 
...                                    mass_matrix=comb_implicit_mat,
...                                    alg_con=alg_con_full,
...                                    coord_idxs=(0, 1))
>>> symsystem3 = system.SymbolicSystem(states, dyn_implicit_rhs, 
...                                    mass_matrix=dyn_implicit_mat,
...                                    coordinate_derivatives=kin_explicit_rhs,
...                                    alg_con=alg_con, coord_idxs=(0, 1), 
...                                    speed_idxs=(2, 3))
```

Isn't that covered by this code?

moorepants commented:
Ok, cool, must have missed it.
